repo/blob/azure/azure_storage.go

```diff
@@ -0,0 +1,273 @@
+	"gocloud.dev/blob/azureblob"
+)
+
+const (
+	azStorageType   = "azure"
+	writerChunkSize = 1 << 20
+)
+
+type azStorage struct {
+	Options
+
+	ctx context.Context
+
+	bucket *gblob.Bucket
+
+	downloadThrottler *iothrottler.IOThrottlerPool
+	uploadThrottler   *iothrottler.IOThrottlerPool
+}
+
+func (az *azStorage) GetBlob(ctx context.Context, b blob.ID, offset, length int64) ([]byte, error) {
+	attempt := func() (interface{}, error) {
+		// TODO: Support chunked download
+		reader, err := az.bucket.NewReader(ctx, az.getObjectNameString(b), nil)
+		if err != nil {
+			return 0, err
+		}
+
+		defer reader.Close() //nolint:errcheck
+
+		throttled, err := az.downloadThrottler.AddReader(reader)
+		if err != nil {
+			return nil, err
+		}
+
+		b, err := ioutil.ReadAll(throttled)
+		if err != nil {
+			return nil, err
+		}
+
+		if len(b) != int(length) && length > 0 {
+			return nil, errors.Errorf("invalid length, got %v bytes, but expected %v", len(b), length)
+		}
+
+		if length == 0 {
+			return []byte{}, nil
+		}
+
+		return b, nil
+	}
+
+	v, err := exponentialBackoff(fmt.Sprintf("GetBlob(%q,%v,%v)", b, offset, length), attempt)
+	if err != nil {
+		return nil, translateError(err)
+	}
+
+	return v.([]byte), nil
+}
+
+func exponentialBackoff(desc string, att retry.AttemptFunc) (interface{}, error) {
+	return retry.WithExponentialBackoff(desc, att, isRetriableError)
+}
+
+func isRetriableError(err error) bool {
+	if me, ok := err.(azblob.ResponseError); ok {
+		if me.Response() == nil {
+			return false
+		}
+		// retry on server errors, not on client errors
+		return me.Response().StatusCode >= 500
+	}
+
+	return false
+}
+
+func translateError(err error) error {
+	switch gcerrors.Code(err) {
+	case gcerrors.NotFound:
+		return blob.ErrBlobNotFound
+
+	// TODO: Handle other possible cases
+	default:
+		return err
+	}
+	return err
+}
+
+func (az *azStorage) PutBlob(ctx context.Context, b blob.ID, data []byte) error {
+	// TODO: Perform chunked upload
+	ctx, cancel := context.WithCancel(ctx)
+
+	writerOptions := gblob.WriterOptions{
+		//BufferSize:  writerChunkSize,
+		ContentType: "application/x-kopia",
+	}
+
+	writer, err := az.bucket.NewWriter(ctx, az.getObjectNameString(b), &writerOptions)
+	if err != nil {
+		return err
+	}
+
+	_, err = writer.Write(data)
+	if err != nil {
+		// cancel context before closing the writer causes it to abandon the upload.
+		cancel()
+
+		_ = writer.Close() // failing already, ignore the error
+
+		return translateError(err)
+	}
+
+	defer cancel()
+
+	// calling close before cancel() causes it to commit the upload.
+	return translateError(writer.Close())
+}
+
+func (az *azStorage) DeleteBlob(ctx context.Context, b blob.ID) error {
+	attempt := func() (interface{}, error) {
+		return nil, az.bucket.Delete(ctx, az.getObjectNameString(b))
+	}
+
+	_, err := exponentialBackoff(fmt.Sprintf("DeleteBlob(%q)", b), attempt)
+
+	return translateError(err)
+}
+
+// TODO: See if we can use PrefixedBucket instead
```

Comment:
What does this mean?

Reply:
We need to check if we can get rid of `getObjectNameString()` if we use PrefixedBucket() function for creating bucket
https://pkg.go.dev/gocloud.dev/blob?tab=doc#PrefixedBucket

Reply:
Checked. Doesn't work for our use case